docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+# Proposal: Index Advisor
+
+- Author(s): [Zou Huan](https://github.com/zouhuan1215) 
+- Last updated:  Nov. 5th, 2019
+- Discussion at: 
+
+## Abstract
+
+This proposal aims at designing a new DBA tool -- index advisor, which automatically recommends indexes for input workload. Having added these recommend indexes to tables, the overall execution time of input workload is supposed to decrease. Current proposal only considers single-column index recommendation.
+
+
+## Background
+
+Appropriate indexes can speed up query execution a lot. However, in real-world applications, determining proper indexes for a database according to a typical workload purely by hand, is an excessively challenging task. When deciding reasonable indexes for a large-scale database, this task becomes even more tough as one database may contain dozens of tables in which each table may have dozens of columns, whereas in the typical workload thousands of queries are involved in. Fortunately, we already have an excellent tool -- TiDB SQL optimizer that can help us get through this tough task. The main purpose of this proposal is to utilize existing TiDB SQL optimizer to automatically recommend suitable indexes for database given the workload.
+
+To illustrate how can SQL optimizer help us determine suitable indexes, here is an intuitive example: 
+
+First assume that all indexes on table `persons` are available. Then comes a query like `SELECT name FROM persons WHERE age > 18`. When responding to this query, the SQL optimizer is more likely to output an execution plan which does *index_scan* upon column `age`. The point here is: **by observing the execution plan output by SQL optimizer, we will know which indexes are favored by input queries.** 
+
+## Proposal
+
+### Input & Output of Index Advisor
+
+The input of index advisor are: `workload`, `parameters`, `statistics`; the output are: `recommend index set`, ` both execution plans before and after adding recommended indexes`. Following shows the details:  
+
+> Input
+1. **Workload**
+
+`workload` can be extracted from *slow_query.log*.
+
+2. **Parameters** (OPTIONAL)
+
+Related parameters and their default values are:
+
+- `database-name="test"`: the database for which index advisor recommend indexes
+- `user="test"`: info required by logging in mysql client
+- `password=""`: info required by logging in mysql client
+- `workload=""`: according to which index advisor recommend indexes
+- `max-exectime=0`: maximum running time of index advisor (unit: minute). Its default value is `0` which means no time constraints for running index advisor
+- `max-index-num=10`: maximum number of indexes recommended by index advisor that is allowed to create 
+- `max-idxcol=1`: maximum number of columns in recommended composite indexes. This proposal set `max-idxcol` as 1
+- `sample-idxstats=false`: whether index advisor is allowed to sample some rows to obtain statistics of composite indexes
+
+3. **Statistics** (OPTIONAL) 
```

Comment:
And it is not needed since we decided to just do it inside tidb.

Reply:
Which one is unnecessary?

Reply:
Statistics. It is not needed in the input session.